script: add skeleton implementation of `FontFace` API

mukilan opened the pull request:
This patch implements the `FontFace` interface, but with some caveats

1. The interface is only exposed on `Window`. Support for Workers will
   be handled in the future.
2. The concept of `css-connected` `FontFace` is not implemented, so
   `@font-face` rules in stylesheets will not be represented in the DOM.
3. The constructor only supports using `url()` strings as source
   and `ArrayBuffer` and `ArrayBufferView` are not supported yet.

A skeleton implementation of the `load` method of `FontFaceSet` is also
implemented in this patch. The intention is to support some web pages
that don't load without this method.

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan committed:
script: add skeleton implementation of `FontFace` API

This patch implements the `FontFace` interface, but with some caveats

1. The interface is only exposed on `Window`. Support for Workers will
   be handled in the future.
2. The concept of `css-connected` `FontFace` is not implemented, so
   `@font-face` rules in stylesheets will not be represented in the DOM.
3. The constructor only supports using `url()` strings as source
   and `ArrayBuffer` and `ArrayBufferView` are not supported yet.

A skeleton implementation of the `load` method of `FontFaceSet` is also
implemented in this patch. The intention is to support some web pages
that don't load without this method.

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

mukilan commented on components/script/dom/fontface.rs:
```diff
@@ -0,0 +1,559 @@
+        );
+
+        let Some(ref font_face_rule_data) = parsed_descriptors else {
+            // Step 1. If any of them fail to parse correctly, reject font face’s
```
All the individual steps are part of the same step in the [spec](https://drafts.csswg.org/css-font-loading/#font-face-constructor). 

I wasn't sure what the convention was in servo codebase for such a case. Do we just leave out the step number?

simonwuelker replied:
Oh I see, thanks.
Yeah I usually leave the step number out.

mrobinson commented on components/fonts/font_context.rs:
```diff
@@ -357,15 +347,104 @@ impl FontContext {
     core_resource_thread: CoreResourceThread,
-    local_fonts: Arc<HashMap<Atom, Option<FontTemplateRef>>>,
-    pub(crate) stylesheet: DocumentStyleSheet,
+    local_fonts: HashMap<Atom, Option<FontTemplateRef>>,
```
Could this still be an `Arc` or `Rc` to avoid copying the data?

mukilan replied:
This change is actually avoiding the copy entirely - before this change `WebFontDownloadState` was `Clone` and we were cloning it for each instance of the download, but with this change `WebFontDownloadState` is a move type, so it is created once and ownership is transferred between the `RemoteWebFontDownloader` instances.

mrobinson commented on components/fonts/font_context.rs:
```diff
@@ -600,12 +716,26 @@ impl FontContext {
 }
 
+pub(crate) enum WebFontLoadInitiator {
+    Stylesheet(DocumentStyleSheet, WebFontLoadFinishedCallback),
+    Script(Box<dyn FnOnce(LowercaseFontFamilyName, Option<FontTemplate>) + Send>),
```
Might be worth renaming `WebFontLoadFinishedCallback` to `StylesheetWebFontLoadFinishedCallback` and then making a type alias called `ScritWebFontFinishedCallback`.

```suggestion
    Stylesheet(DocumentStyleSheet, WebFontLoadFinishedCallback),
    Script(Box<dyn FnOnce(LowercaseFontFamilyName, Option<FontTemplate>) + Send>),
```

mrobinson commented on components/fonts/font_context.rs:
```diff
@@ -653,40 +781,44 @@ impl RemoteWebFontDownloader {
-                        .font_context
-                        .process_next_web_font_source(state.clone()),
+                    DownloaderResponseResult::Failure => {
+                        let state = downloader.take_state();
+                        font_context.process_next_web_font_source(state)
```
```suggestion
                        font_context.process_next_web_font_source(downloader.take_state())
```

mrobinson commented on components/fonts/font_context.rs:
```diff
@@ -653,40 +781,44 @@ impl RemoteWebFontDownloader {
-                            downloader
-                                .font_context
-                                .process_next_web_font_source(state.clone())
+                        if !downloader.process_downloaded_font_and_signal_completion() {
+                            let state = downloader.take_state();
+                            font_context.process_next_web_font_source(state)
```
```suggestion
                        if !downloader.process_downloaded_font_and_signal_completion() {
                            font_context.process_next_web_font_source(downloader.take_state())
```

mrobinson commented on components/fonts/font_context.rs:
```diff
@@ -653,40 +781,44 @@ impl RemoteWebFontDownloader {
-        trace!(
-            "@font-face {} data={:?}",
+        let font_data = std::mem::take(&mut self.response_data);
+        debug!(
```
```suggestion
        trace!(
```

mrobinson commented on components/fonts/font_context.rs:
```diff
@@ -653,40 +781,44 @@ impl RemoteWebFontDownloader {
+            "downloaded @font-face {} data={:?}",
             self.web_font_family_name,
-            font_data
+            font_data.len()
```
Maybe this change is accidental?

mukilan replied:
I noticed that this was actually dumping the data as a vector of bytes to the console, making the log unnecessarily huge and unreadable. Is it ever useful to log the raw bytes of the font?

mrobinson commented on components/fonts/font_context.rs:
```diff
@@ -805,7 +935,7 @@ impl Hash for FontGroupCacheKey {
 }
 
-fn is_supported_web_font_source(source: &&Source) -> bool {
+pub fn is_supported_web_font_source(source: &&Source) -> bool {
```
I don't think this change is necessary?

mrobinson replied:
Okay. Let's make this a static method on `FontContext` then.

mrobinson commented on components/script/dom/fontface.rs:
```diff
@@ -0,0 +1,568 @@
+    ($self:ident, $field: ident, $value: ident) => ({
+        let mut new_descriptors = $self.descriptors.borrow().clone();
+        new_descriptors.$field = $value;
+        let global = $self.global();
+        let parse_result = parse_font_face_descriptors(
+                &global,
+                &$self.family_name.borrow(),
+                None,
+                &new_descriptors
+        );
+        if let Some(ref parsed_font_face_rule) = parse_result {
+            *$self.descriptors.borrow_mut() = serialize_parsed_descriptors(parsed_font_face_rule);
+            Ok(())
+        } else {
+            Err(Error::Syntax)
+        }
```
Could this part of the macro be turned into a helper method? Then each setter would just need to make new descriptor, set the field, and call the helper. That would allow more code reuse and avoid the need for metaprogramming, I think.

mrobinson commented on components/script/dom/fontface.rs:
```diff
@@ -0,0 +1,568 @@
+        debug_assert_eq!(self.status.get(), FontFaceLoadStatus::Unloaded);
+
+        let trusted = Trusted::new(self);
+        let task_source = self
```
If you move `let global = self.global()` above this line you can avoid calling `self.global()` here.

mrobinson commented on components/script/dom/fontface.rs:
```diff
@@ -0,0 +1,568 @@
+        // in [CSS-FONTS-3], as if it was the value of a @font-face rule’s src descriptor.
+        // TODO: FontFaceSet is not supported on Workers yet. The `as_window` call below should be
+        // replaced when we do support it.
+        self.global().as_window().load_web_font_for_script(
```
```suggestion
        global.as_window().load_web_font_for_script(
```

mrobinson commented on components/script/dom/fontfaceset.rs:
```diff
@@ -44,9 +52,20 @@ impl FontFaceSet {
 
+    pub(super) fn handle_font_face_load_completed(&self, font_face: &FontFace) {
+        if font_face.loaded() {
+            if let Some(window) = DomRoot::downcast::<Window>(self.global()) {
+                let (family_name, template) = font_face
+                    .template()
+                    .expect("A loaded web font should have a template");
+                window.add_web_font_to_global_font_context(family_name, template);
+                window.Document().dirty_all_nodes();
+            }
```
Nit: I would use an early return for this.

mrobinson commented on components/script/dom/fontfaceset.rs:
```diff
@@ -44,9 +52,20 @@ impl FontFaceSet {
     }
 
+    pub(super) fn handle_font_face_load_completed(&self, font_face: &FontFace) {
+        if font_face.loaded() {
```
Can you move this check to the `Add` method? It's a little weird this can be called for unloaded fonts. Then this could be an assertion.

mukilan replied:
Eventually this method would handle unloaded fonts as well, to replace its promise and reset the status etc. Maybe I can just rename this to `handle_font_face_status_change`?

mrobinson reviewed (changes requested):
Looking good! Just a few suggestions here:

mukilan commented:
@mrobinson Thanks for the review! I think I've addressed all the comments. There are two tests that are failing intermittently and they don't seem to be caused directly by changes in this PR, but maybe indirectly now that FontFace load() works. I've created intermittent issue for these two - #35520 & #35521.

mrobinson commented:
I think it might be good to do some work to see why these tests are now becoming flaky as, in general, we should be moving toward less flakiness when adding API. It could point to some issue with your code or, at least, let you know what the next part to work on is.

mukilan commented:
> I think it might be good to do some work to see why these tests are now becoming flaky as, in general, we should be moving toward less flakiness when adding API. It could point to some issue with your code or, at least, let you know what the next part to work on is.

I've updated the expectations for the two failing tests to be account for the intermittent results. As discussed, they are flaky because the rest of the font stack is not complete yet -  1) lack of support for CSS-Connected FontFace objects from stylesheets and having a proper implementation of `load()` that tracks their load progress and 2) there is no support for the font's display descriptor (all major engines are failing this test https://wpt.fyi/results/css/css-fonts/font-display?label=experimental&label=master&aligned).

Should I close the two intermittent new issues?

mrobinson commented on tests/wpt/meta/css/css-cascade/layer-font-face-override.html.ini:
```diff
@@ -1,9 +1,9 @@
 [layer-font-face-override.html]
```
Please also open a bug describing this problem so that we can track fixing the promise returned by `FontFaceSet.load(...)`. In addition, there is syntax for linking to the bug in the ini file which we use in other places. I think it makes sense to do that here.

mrobinson reviewed (approved):
Yeah, I think it makes sense to close the intermittent bugs or rework them to track the problems that cause the intermittent tests. That will allow us to track what is causing the issues.
